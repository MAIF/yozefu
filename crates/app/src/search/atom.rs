use async_trait::async_trait;
use lib::search::{atom::Atom, filter::Filter, offset::FromOffset};

use super::{Search, SearchContext};

#[async_trait]
impl Search for Atom {
    fn offset(&self) -> Option<FromOffset> {
        match self {
            Atom::Symbol(_) => None,
            Atom::Compare(c) => c.offset(),
            Atom::Parenthesis(c) => c.offset(),
            Atom::Filter(_) => None,
        }
    }

    fn matches(&self, context: &SearchContext) -> bool {
        match self {
            Atom::Symbol(_) => false,
            Atom::Compare(e) => e.matches(context),
            Atom::Parenthesis(e) => e.matches(context),
            Atom::Filter(f) => f.matches(context),
        }
    }

    fn filters(&self) -> Vec<Filter> {
        match self {
            Atom::Symbol(_) => vec![],
            Atom::Compare(e) => e.filters(),
            Atom::Parenthesis(e) => e.filters(),
            Atom::Filter(f) => vec![f.clone()],
        }
    }
}
