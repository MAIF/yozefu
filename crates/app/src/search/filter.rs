use std::{
    collections::HashMap,
    sync::{LazyLock, Mutex},
};

use extism::{convert::Json, Plugin};
use itertools::Itertools;
use lib::{
    search::filter::{Filter, FilterInput},
    FilterResult,
};

use super::{Search, SearchContext};

// This is evil, TODO context
pub static CACHED_FILTERS: LazyLock<Mutex<HashMap<String, Plugin>>> =
    LazyLock::new(|| HashMap::new().into());

pub const MATCHES_FUNCTION_NAME: &str = "matches";
pub const PARSE_PARAMETERS_FUNCTION_NAME: &str = "parse_parameters";

impl Search for Filter {
    fn matches(&self, context: &SearchContext) -> bool {
        let mut filters = CACHED_FILTERS.lock().unwrap();
        let plugin = &mut filters.get_mut(&self.name).unwrap();
        let input = FilterInput {
            record: context.record.clone(),
            params: self.parameters.iter().map(|e| e.json()).collect_vec(),
        };

        match plugin
            .call::<String, Json<FilterResult>>(
                MATCHES_FUNCTION_NAME,
                serde_json::to_string(&input).unwrap(),
            )
            .map(|e| e.0)
        {
            Ok(res) => res.r#match,
            Err(_e) => true,
        }
    }

    fn filters(&self) -> Vec<Filter> {
        vec![]
    }
}
