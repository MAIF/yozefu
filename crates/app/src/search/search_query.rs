use lib::{
    search::{filter::Filter, offset::FromOffset},
    SearchQuery,
};

use super::{Search, SearchContext};

impl Search for SearchQuery {
    fn offset(&self) -> Option<FromOffset> {
        self.from.clone().or(self.expression.offset())
    }

    fn matches(&self, context: &SearchContext) -> bool {
        self.expression.matches(context)
    }

    fn filters(&self) -> Vec<Filter> {
        self.expression.filters()
    }
}
