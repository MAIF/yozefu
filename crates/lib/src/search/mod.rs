//! This module defines the the parsing functions of search query.
//! The grammar of the syntax is the following:
//!
//! ```bnf
//! search-query      ::= clause+
//! clause            ::= or-expression | limit-clause | from-clause | order-clause
//! or-expression     ::= And-expression | and-expression 'or' and-expression
//! and-expression    ::= atom | atom 'and' atom
//! term              ::= atom | '!' atom
//! atom              ::= comparison  | filter | '(' expression ')'
//! number-symbol     ::= 'offset' | 'partition' | 'size'
//! string-symbol     ::= 'topic' | 'key' | 'timestamp' | 'value'
//! symbol            ::= number-symbol | string-symbol
//! comparison        ::= number-comparison | string-comparison | time-comparison
//! number-comparison ::= number-symbol number-operator number
//! string-comparison ::= string-symbol string-operator string
//! time-comparison   ::= 'between' string 'and' string
//! number-operator   ::=  '==' | '!=' | '>' | '<' | '>=' | '<='
//! string-operator   ::= 'starts with' | '==' | '!=' | '=~' | 'contains' | 'contain' | 'includes' | 'include'
//! filter            ::= .+ '('filter-parameters')'
//! filter-parameter  ::= string | number
//! filter-parameters ::= filter-parameter  (',' filter-parameter)*
//! limit-clause      ::= 'limit' number
//! order-clause      ::= 'order by' symbol order-keyword
//! order-keyword     ::= 'asc' | 'desc'
//! from-clause       ::= 'from' offset
//! offset            ::= 'beginning' | 'begin' | 'end' | 'end' '-' number | string | number
//! number            ::= [0-9_]+
//! string            ::= '"' [^"]+ '"' | "'" [^']+ "'"
//! ```
//! You can use <https://www.bottlecaps.de/rr/ui> to visualize it.

#[cfg(feature = "native")]
pub mod atom;
#[cfg(feature = "native")]
pub mod clause;
#[cfg(feature = "native")]
pub mod expression;
#[cfg(feature = "native")]
pub mod filter;
#[cfg(feature = "native")]
pub mod number;
#[cfg(feature = "native")]
pub mod offset;
#[cfg(feature = "native")]
pub mod order;
#[cfg(feature = "native")]
pub mod search_query;
#[cfg(feature = "native")]
pub mod string;
#[cfg(feature = "native")]
pub mod symbol;
#[cfg(feature = "native")]
pub mod term;
#[cfg(feature = "native")]
pub mod timestamp;
#[cfg(feature = "native")]
pub mod wsi;

pub mod compare;

#[cfg(feature = "native")]
pub use order::Order;
#[cfg(feature = "native")]
pub use order::OrderBy;
#[cfg(feature = "native")]
pub use search_query::parse_search_query;
#[cfg(feature = "native")]
pub use search_query::SearchQuery;
use serde::Deserialize;
use serde::Serialize;

#[cfg(test)]
pub mod expression_test;
#[cfg(test)]
pub mod number_test;
#[cfg(test)]
pub mod offset_test;

#[derive(Debug, PartialEq, Clone, Default, Deserialize, Serialize)]
pub struct FilterResult {
    pub r#match: bool,
}

impl FilterResult {
    pub fn new(r#match: bool) -> Self {
        Self { r#match }
    }
}

impl From<bool> for FilterResult {
    fn from(r#match: bool) -> Self {
        Self { r#match }
    }
}
