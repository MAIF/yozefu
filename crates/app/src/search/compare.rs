use async_trait::async_trait;
use lib::{
    kafka::Comparable,
    search::{
        compare::{CompareExpression, NumberOperator, StringOperator},
        filter::Filter,
        offset::FromOffset,
    },
};

use crate::search::Search;

use super::SearchContext;

#[async_trait]
impl Search for CompareExpression {
    fn offset(&self) -> Option<FromOffset> {
        match self {
            CompareExpression::Offset(NumberOperator::Equal, e) => Some(FromOffset::Offset(*e)),
            CompareExpression::Offset(NumberOperator::GreaterOrEqual, e) => {
                Some(FromOffset::Offset(*e))
            }
            CompareExpression::Offset(NumberOperator::GreaterThan, e) => {
                Some(FromOffset::Offset(*e + 1))
            }
            CompareExpression::OffsetTail(e) => Some(FromOffset::OffsetTail(*e)),
            CompareExpression::Timestamp(op, e) => match op {
                NumberOperator::GreaterThan => Some(FromOffset::Timestamp(e.timestamp_millis())),
                NumberOperator::GreaterOrEqual => {
                    Some(FromOffset::Timestamp(e.timestamp_millis() - 1000))
                }
                NumberOperator::Equal => Some(FromOffset::Timestamp(e.timestamp_millis() - 1000)),
                NumberOperator::LowerThan => None,
                NumberOperator::LowerOrEqual => None,
                _ => None,
            },
            _ => None,
        }
    }

    fn matches(&self, context: &SearchContext) -> bool {
        let record = context.record;
        match self {
            CompareExpression::OffsetTail(_) => true,
            CompareExpression::Partition(op, p) => match op {
                &NumberOperator::GreaterThan => record.partition > *p,
                NumberOperator::GreaterOrEqual => record.partition >= *p,
                NumberOperator::LowerThan => record.partition < *p,
                NumberOperator::LowerOrEqual => record.partition <= *p,
                NumberOperator::Equal => record.partition == *p,
                NumberOperator::NotEqual => record.partition != *p,
            },
            CompareExpression::Offset(op, p) => match op {
                NumberOperator::GreaterThan => record.offset > *p,
                NumberOperator::GreaterOrEqual => record.offset >= *p,
                NumberOperator::LowerThan => record.offset < *p,
                NumberOperator::LowerOrEqual => record.offset <= *p,
                NumberOperator::Equal => record.offset == *p,
                NumberOperator::NotEqual => record.offset != *p,
            },
            CompareExpression::Topic(op, t) => match op {
                StringOperator::Equal => record.topic == *t,
                StringOperator::NotEqual => record.topic != *t,
                StringOperator::Contain => record.topic.contains(t),
                StringOperator::StartWith => record.topic.starts_with(t),
            },
            CompareExpression::Size(op, s) => match op {
                NumberOperator::GreaterThan => record.size > *s as usize,
                NumberOperator::GreaterOrEqual => record.size >= *s as usize,
                NumberOperator::LowerThan => record.size < *s as usize,
                NumberOperator::LowerOrEqual => record.size <= *s as usize,
                NumberOperator::Equal => record.size == *s as usize,
                NumberOperator::NotEqual => record.size != *s as usize,
            },
            CompareExpression::Key(op, t) => record.key.compare(&None, op, t),
            CompareExpression::Value(left, op, t) => record.value.compare(left, op, t),
            CompareExpression::Header(left, op, t) => {
                let headers = &record.headers;
                let header = headers.get(left);
                if header.is_none() {
                    return false;
                }
                let header = header.unwrap();
                match op {
                    StringOperator::Contain => header.contains(t),
                    StringOperator::Equal => header == t,
                    StringOperator::StartWith => header.starts_with(t),
                    StringOperator::NotEqual => header != t,
                }
            }
            CompareExpression::Timestamp(op, t) => {
                let ts = record.timestamp_as_local_date_time().unwrap();
                match op {
                    NumberOperator::GreaterThan => ts > *t,
                    NumberOperator::GreaterOrEqual => ts >= *t,
                    NumberOperator::LowerThan => ts < *t,
                    NumberOperator::LowerOrEqual => ts <= *t,
                    NumberOperator::Equal => ts == *t,
                    NumberOperator::NotEqual => ts != *t,
                }
            }
            CompareExpression::TimestampBetween(from, to) => {
                let ts = record.timestamp_as_local_date_time().unwrap();
                from <= &ts && &ts <= to
            }
        }
    }

    fn filters(&self) -> Vec<Filter> {
        vec![]
    }
}
