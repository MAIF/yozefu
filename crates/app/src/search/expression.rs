use lib::search::{
    expression::{AndExpression, OrExpression},
    filter::Filter,
    offset::FromOffset,
};

use super::{Search, SearchContext};

impl Search for AndExpression {
    fn offset(&self) -> Option<FromOffset> {
        match self {
            Self::AndTerm(t) => t.offset(),
            Self::AndExpression(v) => {
                for vv in v {
                    let o = vv.offset();
                    if o.is_some() {
                        return o;
                    }
                }
                None
            }
        }
    }

    fn matches(&self, context: &SearchContext) -> bool {
        let record = context;
        match self {
            Self::AndTerm(t) => t.matches(record),
            Self::AndExpression(e) => {
                for ee in e {
                    if !ee.matches(record) {
                        return false;
                    }
                }
                true
            }
        }
    }

    fn filters(&self) -> Vec<Filter> {
        match self {
            AndExpression::AndTerm(term) => term.filters(),
            AndExpression::AndExpression(vec) => vec.iter().flat_map(|t| t.filters()).collect(),
        }
    }
}

impl Search for OrExpression {
    fn offset(&self) -> Option<FromOffset> {
        match self {
            Self::OrTerm(t) => t.offset(),
            Self::OrExpression(_) => None,
        }
    }

    fn matches(&self, context: &SearchContext) -> bool {
        match self {
            Self::OrTerm(t) => t.matches(context),
            Self::OrExpression(e) => {
                if e.is_empty() {
                    return true;
                }
                for ee in e {
                    if ee.matches(context) {
                        return true;
                    }
                }
                false
            }
        }
    }

    fn filters(&self) -> Vec<Filter> {
        match self {
            OrExpression::OrTerm(and_expression) => and_expression.filters(),
            OrExpression::OrExpression(vec) => vec.iter().flat_map(|e| e.filters()).collect(),
        }
    }
}
