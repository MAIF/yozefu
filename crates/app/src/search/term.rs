use lib::search::{filter::Filter, offset::FromOffset, term::Term};

use super::{Search, SearchContext};

impl Search for Term {
    fn offset(&self) -> Option<FromOffset> {
        match self {
            Term::Not(_) => None,
            Term::Atom(a) => a.offset(),
        }
    }

    fn matches(&self, context: &SearchContext) -> bool {
        match self {
            Term::Not(a) => !a.matches(context),
            Term::Atom(a) => a.matches(context),
        }
    }

    fn filters(&self) -> Vec<Filter> {
        match self {
            Term::Not(atom) => atom.filters(),
            Term::Atom(atom) => atom.filters(),
        }
    }
}
