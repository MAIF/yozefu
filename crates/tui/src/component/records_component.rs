//! Component showing in real time incoming kafka records.
use app::search::ValidSearchQuery;
use copypasta::{ClipboardContext, ClipboardProvider};
use crossterm::event::{KeyCode, KeyEvent, KeyModifiers};
use itertools::Itertools;
use lib::ExportedKafkaRecord;
use ratatui::{
    Frame,
    layout::{Alignment, Constraint, Rect},
    style::{Style, Stylize},
    text::{Span, Text},
    widgets::{Block, BorderType, Borders, Cell, Row, Table, TableState},
};
use thousands::Separable;
use throbber_widgets_tui::ThrobberState;
use tokio::sync::mpsc::UnboundedSender;
use tokio::sync::watch::Receiver;

use crate::{Action, action::Notification, error::TuiError, records_buffer::BufferAction};

use super::{Component, ComponentName, ConcurrentRecordsBuffer, Shortcut, State};

pub(crate) struct RecordsComponent<'a> {
    records: &'a ConcurrentRecordsBuffer,
    state: TableState,
    status: ThrobberState,
    search_query: ValidSearchQuery,
    consuming: bool,
    count: (usize, usize, usize),
    follow: bool,
    action_tx: Option<UnboundedSender<Action>>,
    buffer_tx: Receiver<BufferAction>,
    selected_topics: usize,
    key_events_buffer: Vec<KeyEvent>,
}

impl<'a> RecordsComponent<'a> {
    pub fn new(records: &'a ConcurrentRecordsBuffer) -> Self {
        let buffer_tx = records.lock().map(|e| e.channels.clone().1).ok().unwrap();

        Self {
            records,
            state: Default::default(),
            status: Default::default(),
            search_query: Default::default(),
            consuming: Default::default(),
            count: Default::default(),
            follow: Default::default(),
            action_tx: Default::default(),
            buffer_tx,
            selected_topics: Default::default(),
            key_events_buffer: Default::default(),
        }
    }

    fn buffer_is_empty(&self) -> bool {
        self.count.2 == 0
    }

    fn buffer_len(&self) -> usize {
        self.count.2
    }

    fn next(&mut self) {
        if self.buffer_is_empty() {
            self.state.select(None);
            return;
        }
        let i = match self.state.selected() {
            Some(i) => {
                if i >= self.buffer_len() - 1 {
                    i
                } else {
                    i + 1
                }
            }
            None => 0,
        };
        self.state.select(Some(i));
    }

    fn shorten_topic(topic: &str) -> String {
        let t = topic.replace('_', "");
        let parts = t.split('.');
        let e = parts.map(|e| e.chars().next().unwrap_or('_')).join(".");
        e
    }

    fn show_details(&mut self) -> Result<(), TuiError> {
        if self.state.selected().is_some() {
            self.action_tx
                .as_ref()
                .unwrap()
                .send(Action::NewView(ComponentName::RecordDetails))?;
            self.set_event_dialog()?
        }
        Ok(())
    }

    fn set_event_dialog(&mut self) -> Result<(), TuiError> {
        if let Some(s) = self.state.selected() {
            let record = self.records.lock().unwrap().get(s).unwrap().clone();
            self.action_tx
                .as_ref()
                .unwrap()
                .send(Action::ShowRecord(record))?;
        }
        Ok(())
    }

    fn previous(&mut self) {
        if self.buffer_is_empty() {
            self.state.select(None);
            return;
        }
        let i = match self.state.selected() {
            Some(i) => {
                if i == 0 {
                    0
                } else {
                    i - 1
                }
            }
            None => 0,
        };
        self.state.select(Some(i));
    }

    fn first(&mut self) {
        match self.buffer_is_empty() {
            true => self.state.select(None),
            false => self.state.select(Some(0)),
        }
    }

    fn last(&mut self) {
        match self.buffer_is_empty() {
            true => self.state.select(None),
            false => self.state.select(Some(self.buffer_len() - 1)),
        }
    }

    pub fn on_new_record(&mut self, count: (usize, usize, usize)) -> Result<(), TuiError> {
        self.count = count;
        let length = count.2;
        let empty_buffer = length == 0;
        if self.follow && !empty_buffer {
            self.state.select(Some(length - 1));
        }
        if self.state.selected().is_none() && !empty_buffer {
            self.state.select(Some(0));
        }
        if let Some(s) = self.state.selected() {
            if s >= length {
                let ii = match empty_buffer {
                    true => 0,
                    false => length - 1,
                };
                self.state.select(Some(ii));
            }
        }
        Ok(())
    }

    fn buffer_key_event(&mut self, e: KeyEvent) -> Result<(), TuiError> {
        self.key_events_buffer.push(e);
        if self.key_events_buffer.len() < 2 {
            return Ok(());
        }
        self.key_events_buffer.dedup();
        if self.key_events_buffer.len() == 1 {
            match self.key_events_buffer.first().unwrap().code {
                KeyCode::Char('g') => {
                    self.follow(false)?;
                    self.first();
                }
                KeyCode::Char('G') => {
                    self.follow(false)?;
                    self.last();
                }
                _ => (),
            }
        }
        self.key_events_buffer.clear();
        Ok(())
    }

    fn follow(&mut self, follow: bool) -> Result<(), TuiError> {
        self.follow = follow;
        if self.follow {
            self.state.select(match self.buffer_len() {
                0 => None,
                i => Some(i - 1),
            });
        }
        self.action_tx
            .as_ref()
            .unwrap()
            .send(Action::RefreshShortcuts)?;
        Ok(())
    }

    fn truncate_value(value: &str, rect: &Rect) -> String {
        let split_at = (rect.width.checked_sub(70)).unwrap_or(3) as usize;
        match value.len() > split_at {
            true => value.chars().take(split_at).collect(),
            false => value.to_string(),
        }
    }
}

impl Component for RecordsComponent<'_> {
    fn register_action_handler(&mut self, tx: UnboundedSender<Action>) {
        self.action_tx = Some(tx.clone());
    }

    fn id(&self) -> ComponentName {
        ComponentName::Records
    }

    fn handle_key_events(&mut self, key: KeyEvent) -> Result<Option<Action>, TuiError> {
        match key.code {
            KeyCode::Char('c') => {
                if let Some(s) = self.state.selected() {
                    let r = self.records.lock().unwrap();
                    let record = r.get(s).unwrap();
                    let mut ctx = ClipboardContext::new().unwrap();
                    let exported_record: ExportedKafkaRecord = record.into();
                    self.action_tx.as_ref().unwrap().send(Action::Notification(
                        Notification::new(log::Level::Info, "Copied to clipboard".to_string()),
                    ))?;
                    ctx.set_contents(serde_json::to_string_pretty(&exported_record)?)
                        .unwrap();
                }
            }
            KeyCode::Char('f') => self.follow(!self.follow)?,
            KeyCode::Char('v') | KeyCode::Enter => {
                self.show_details()?;
            }
            KeyCode::Char('e') if key.modifiers.contains(KeyModifiers::CONTROL) => {
                let records = self.records.lock().unwrap();
                for record in records.iter() {
                    self.action_tx
                        .as_ref()
                        .unwrap()
                        .send(Action::Export(record.clone()))?;
                }
            }
            KeyCode::Char('e') => {
                if let Some(s) = self.state.selected() {
                    let r = self.records.lock().unwrap();
                    let record = r.get(s).unwrap();
                    self.action_tx
                        .as_ref()
                        .unwrap()
                        .send(Action::Export(record.clone()))?;
                }
            }
            KeyCode::Char('o') => {
                if let Some(s) = self.state.selected() {
                    let r = self.records.lock().unwrap();
                    let record = r.get(s).unwrap();
                    self.action_tx
                        .as_ref()
                        .unwrap()
                        .send(Action::Open(record.clone()))?;
                }
            }
            KeyCode::Char('[') => {
                self.follow(false)?;
                self.first();
            }
            KeyCode::Char(']') => {
                self.follow(false)?;
                self.last();
            }
            KeyCode::Down => {
                self.follow(false)?;
                self.next();
                self.set_event_dialog()?;
            }
            KeyCode::Up => {
                self.follow(false)?;
                self.previous();
                self.set_event_dialog()?;
            }
            KeyCode::Char('g') | KeyCode::Char('G') => self.buffer_key_event(key)?,
            _ => (),
        }
        Ok(None)
    }

    fn update(&mut self, action: Action) -> Result<Option<Action>, TuiError> {
        let mut a = self.buffer_tx.clone();
        let BufferAction::Count(count) = *a.borrow_and_update();
        let _ = self.on_new_record(count);
        match action.clone() {
            Action::NewConsumer() => {
                self.count = (0, 0, 0);
            }
            Action::Tick => self.status.calc_next(),
            Action::SelectedTopics(topics) => self.selected_topics = topics.len(),
            Action::Consuming => self.consuming = true,
            Action::StopConsuming() => {
                self.consuming = false;
                self.count = (0, 0, 0);
            }
            Action::Search(search_query) => {
                self.state.select(None);
                self.search_query = search_query;
            }
            _ => (),
        };
        Ok(None)
    }

    fn draw(&mut self, f: &mut Frame<'_>, rect: Rect, state: &State) -> Result<(), TuiError> {
        let focused = state.is_focused(self.id());
        let block = Block::default()
            .borders(Borders::ALL)
            .border_type(BorderType::Rounded)
            .title(" Records ");

        let block = self.make_block_focused_with_state(state, block);

        let normal_style = Style::default();
        let header_cells = vec![
            Cell::new(Text::from("Timestamp")),
            Cell::new(Text::from("Offset").alignment(Alignment::Right)),
            Cell::new(Text::from("Partition").alignment(Alignment::Right)),
            Cell::new(Text::from("Topic").alignment(Alignment::Right)),
            Cell::new(Text::from("Key").alignment(Alignment::Right)),
            Cell::new(Text::from("Value")),
        ];
        let header = Row::new(header_cells)
            .style(normal_style)
            .height(1)
            .bottom_margin(1);
        let r = self.records.lock().unwrap();

        // TODO render only records in the viewport
        let rows = r.iter().enumerate().map(|(index, item)| {
            if let Some(s) = self.state.selected() {
                let is_visible = (s + rect.height as usize) > index
                    && s.saturating_sub(rect.height as usize) <= index;
                if !is_visible {
                    return Row::new(Vec::<Cell>::new()).height(1_u16);
                }
            }
            let cells = vec![
                Cell::new(Text::from(
                    item.timestamp_as_local_date_time()
                        .map(|e| e.to_rfc3339_opts(chrono::SecondsFormat::Millis, false))
                        .unwrap_or("".to_string()),
                )),
                Cell::new(Text::from(item.offset.to_string()).alignment(Alignment::Right)),
                Cell::new(Text::from(item.partition.to_string()).alignment(Alignment::Right)),
                Cell::new(Text::from(Self::shorten_topic(&item.topic)).alignment(Alignment::Right)),
                Cell::new(Text::from(item.key_as_string.to_string()).alignment(Alignment::Right)),
                Cell::new(Text::from(Self::truncate_value(
                    &item.value_as_string,
                    &rect,
                ))),
            ];
            Row::new(cells).height(1_u16)
        });
        let table = Table::new(
            rows,
            [
                Constraint::Min(30),
                Constraint::Min(10),
                Constraint::Min(10),
                Constraint::Min(11),
                Constraint::Min(12),
                Constraint::Percentage(100),
            ],
        )
        .header(header)
        .row_highlight_style(match focused {
            true => Style::default()
                .bg(state.theme.bg_focused_selected)
                .fg(state.theme.fg_focused_selected)
                .bold(),
            false => Style::default()
                .bg(state.theme.bg_unfocused_selected)
                .fg(state.theme.fg_unfocused_selected),
        });

        let metrics = Span::styled(
            format!(
                " {} / {} ",
                self.count.0.separate_with_underscores(),
                self.count.1.separate_with_underscores()
            ),
            Style::default(),
        );
        let inner = block.inner(rect);
        f.render_widget(block, rect);

        f.render_stateful_widget(table, inner, &mut self.state);
        let metrics_area = Rect::new(
            inner
                .right()
                .checked_sub(u16::try_from(metrics.width())?)
                .unwrap_or(1000)
                .checked_sub(10)
                .unwrap_or(1000),
            inner.y,
            metrics.width() as u16,
            1,
        );

        if self.consuming && self.count.1 != 0 {
            f.render_widget(metrics, metrics_area);
        }
        if self.consuming {
            let simple = throbber_widgets_tui::Throbber::default();
            let ss = Span::styled(
                " Live    ",
                Style::default()
                    .fg(state.theme.white)
                    .bg(state.theme.orange),
            )
            .bold();
            f.render_widget(
                ss,
                Rect::new(inner.right().saturating_sub(9), inner.y, 8, 1),
            );
            f.render_stateful_widget(
                simple,
                Rect::new(inner.right().saturating_sub(3), inner.y, 2, 1),
                &mut self.status,
            );
        }
        Ok(())
    }

    fn shortcuts(&self) -> Vec<Shortcut> {
        let shortcuts = vec![
            Shortcut::new("C", "Copy"),
            Shortcut::new("O", "Open"),
            // Shortcut::new("[", "First record"),
            // Shortcut::new("]", "Last record"),
            Shortcut::new("E", "Export"),
            Shortcut::new(
                "F",
                match self.follow {
                    true => "Unfollow",
                    false => "Follow",
                },
            ),
            //Shortcut::new("↑↓", "Scroll"),
        ];

        shortcuts
    }
}
