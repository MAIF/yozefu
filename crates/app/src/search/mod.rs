//! Module implementing the search logic

use std::path::{Path, PathBuf};

use extism::{Manifest, Plugin, Wasm};
use filter::{CACHED_FILTERS, PARSE_PARAMETERS_FUNCTION_NAME};
use itertools::Itertools;
use lib::{
    parse_search_query,
    search::{filter::Filter, offset::FromOffset},
    KafkaRecord, SearchQuery,
};
use log::error;

pub mod atom;
pub mod compare;
pub mod expression;
pub mod filter;
pub mod search_query;
pub mod term;

pub trait Search {
    /// Returns the offset from which the search should start.
    fn offset(&self) -> Option<FromOffset> {
        None
    }
    /// returns `true` if the record matches the search query.
    fn matches(&self, context: &SearchContext) -> bool;

    /// Returns the search filters that are used in the search query.
    fn filters(&self) -> Vec<Filter>;
}

/// Struct that holds the context of the search.
/// It contains the record that is being searched and the loaded search filters.
pub struct SearchContext<'a> {
    pub record: &'a KafkaRecord,
    pub filters_directory: PathBuf,
}

impl SearchContext<'_> {
    pub fn new<'a>(record: &'a KafkaRecord, filters_directory: &'a Path) -> SearchContext<'a> {
        SearchContext {
            record,
            filters_directory: filters_directory.to_path_buf(),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Default)]
pub struct ValidSearchQuery(SearchQuery);

impl ValidSearchQuery {
    pub fn is_empty(&self) -> bool {
        self.0.is_empty()
    }

    pub fn limit(&self) -> Option<usize> {
        self.0.limit
    }

    pub fn query(&self) -> &SearchQuery {
        &self.0
    }
}

impl ValidSearchQuery {
    pub fn from(input: &str, filters_directory: &Path) -> Result<Self, lib::Error> {
        let query = parse_search_query(input).map_err(lib::Error::Search)?.1;
        let filters = query.filters();
        for filter in filters {
            let name = filter.name;
            let path = filters_directory.join(format!("{}.wasm", &name));
            let url = Wasm::file(&path);
            let manifest = Manifest::new([url]);
            let mut filters = CACHED_FILTERS.lock().unwrap();
            if !filters.contains_key(&name) {
                match Plugin::new(manifest, [], true) {
                    Ok(plugin) => filters.insert(name.to_string(), plugin),
                    Err(err) => {
                        error!("No such file '{}': {}", path.display(), err);
                        return Err(lib::Error::Error(format!(
                            "Cannot find search filter '{}'",
                            name
                        )));
                    }
                };
            }
            let params = filter.parameters;
            let wasm_module = &mut filters.get_mut(&name).unwrap();
            if let Err(e) = wasm_module.call::<&str, &str>(
                PARSE_PARAMETERS_FUNCTION_NAME,
                &serde_json::to_string(&params.iter().map(|e| e.json()).collect_vec()).unwrap(),
            ) {
                error!(
                    "Error when calling '{}' from wasm module '{}': {:?}",
                    PARSE_PARAMETERS_FUNCTION_NAME, name, e
                );
                return Err(lib::Error::Error(format!("{}: {e}", &name)));
            };
        }

        Ok(ValidSearchQuery(query))
    }
}

impl Search for ValidSearchQuery {
    /// Returns the offset from which the search should start.
    fn offset(&self) -> Option<FromOffset> {
        self.0.offset()
    }

    fn matches(&self, context: &SearchContext) -> bool {
        self.0.matches(context)
    }

    fn filters(&self) -> Vec<Filter> {
        self.0.filters()
    }
}
