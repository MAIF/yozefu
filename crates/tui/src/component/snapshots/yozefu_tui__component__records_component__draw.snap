---
source: crates/tui/src/component/records_component.rs
expression: terminal.backend()
---
"╭ Records ─────────────────────────────────────────────────────────────────────────────────────────────────────────────╮"
"│Timestamp                             Topic   Offset         Key  Value                                               │"
"│                                                                                                                      │"
"│1970-01-01T01:00:00.000+01:00  movi..ers[0]      314  7f12bd3b..                                                      │"
"╰──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯"
